src/mrms/mrms_grib2ioda.py

```diff
@@ -168,84 +168,112 @@ def main(file_names, output_file):
-        sys.exit()
+    with open(input_file, "rb") as f:
+        try:
+            iid = eccodes.codes_index_new_from_file(input_file, ["level"])
+            levs = eccodes.codes_index_get(iid, "level")
+            for height in levs:
+                heights.append(height)
+        except Exception as e:
+            logging.warning(f"ABORT, failure to open file: {input_file}. MSG: {e}")
+            sys.exit()
 
-    lats = lats.reshape(ni*nj).astype('float')
-    if mask is not None and (len(heights) == 1):
-        lats = lats[mask]
-    lats = lats.tolist()
-    lons = lons.reshape(ni*nj).astype('float')
-    lons[lons > 180.0] = lons - 360.0
-    if mask is not None and (len(heights) == 1):
-        lons = lons[mask]
-    lons = lons.tolist()
+    for height in heights:
+        logging.debug(f"DEBUG: deciphering data at height: {height}")
+        eccodes.codes_index_select(iid, "level", height)
+        gid = eccodes.codes_new_from_index(iid)
+
+        product_id = ''
+        for key in grib_keys:
+            product_id = product_id + str(eccodes.codes_get(gid, key)) + '-'
+        product_id = product_id[:-1]
+        logging.debug(f"DEBUG: the grib file has variable reference: {product_id}")
+
+        '''
+        # Use a mask to remove the missing value points entirely from the dataset.
+        # This is FLAWED because there could be different missing values on different
+        # height levels, or potentially different missing values depending on product.
+        # Currently -999 values are outside radar coverage and -99 values are used
+        # for valid signal below minimum detectable threshold for reflectivity.
+        '''
+
+        mask = None
+        if product_id in mrms_products.keys():
+            obsvar = mrms_products[product_id]
+            d = eccodes.codes_get(gid, "dataDate")
+            t = eccodes.codes_get(gid, "dataTime")
+            if t > 2359:
+                logging.debug(f"DEBUG: dataTime is more than 4 digits, adjusting")
+                t = t/100
+            dt = datetime.strptime(str(d)+str(t), "%Y%m%d%H%M")
+            logging.debug(f"DEBUG: date info: {d} {t}Z")
+
+            ni = eccodes.codes_get(gid, "Ni")
+            nj = eccodes.codes_get(gid, "Nj")
+            logging.debug(f"DEBUG: number of x,y points {ni}, {nj} and level: {height}")
+
+            lats = eccodes.codes_get_array(gid, "latitudes")
+            lons = eccodes.codes_get_array(gid, "longitudes")
+
+            Z = eccodes.codes_get_double_array(gid, 'values')
+
+            Z = Z.reshape(ni*nj).astype('float')
+            if obsvar == 'reflectivity':
+                mask_inside = np.logical_and(Z <= -98.5, Z > -99.5)  # Capture -99 as special
+                Z[mask_inside] = -34.5                               # and reset to -34.5
+                mask = np.logical_and(Z >= min_dbz, Z <= 80)
+                Z = Z[mask]
```

Comment:
If the `-999` and `-99` values are well-known numbers for this dataset, I suggest keeping them as they are (except for `-999` because the information is useless).

I don't know what would be the consequences down the road, but my take would be to have a mask that satisfies the three conditions. Not tested, but something like below:

```suggestion
                mask = np.logical_and(Z >= min_dbz, Z <= 80)
                mask = np.logical_or(Z == -99, mask)
                Z = Z[mask]
```

This would keep `-99` in the dataset and it would allow the user to play with specific filters later.

Reply:
I propose keeping my change of -99 to a value near -35 for physical reasons and matching with long-standing radar reflectivity code introduced in WRF and adopted by MPAS and UFS-GFS/RRFS.

Reply:
Sure!